Docs: Augment Grafana Cloud Migration Assistant docs with troubleshooting, disaster recovery, etc. (#126018)
grafana/grafana: docs/sources/administration/migration-guide/cloud-migration-assistant.md

- Library Panels
- Grafana Alerting resources

> Support for these resources has expanded over recent Grafana releases:

Reviewer: We intentionally removed these notices when this went GA as we maintain versioned docs that call out availability (e.g. https://archive.grafana.com/docs/grafana/v11.5/administration/migration-guide/cloud-migration-assistant/) and none of these versions of Grafana are officially supported anymore - I'd lean towards keeping that information in the archive as we want everything in `latest` to be reflective of the products and feature as they are currently

Author: I'll remove right now. Thanks for the review @jtvdez

- **Grafana 11.2**: Dashboards, folders, and data sources.
- **Grafana 11.5**: All plugins (data sources, app plugins, and panel plugins), library panels, and Grafana Alerting resources.
- **Grafana 12**: The migration assistant is generally available.

## Before you begin

To use the Grafana migration assistant, you need:

> - A self-managed Grafana instance running Grafana 11.2 or later. In Grafana 12 and later, the migration assistant is generally available. In Grafana 11.2 through 11.6, the migration assistant is available in public preview and requires the `onPremToCloudMigrations` [feature toggle](https://grafana.com/docs/grafana/<GRAFANA_VERSION>/setup-grafana/configure-grafana/feature-toggles/) to be enabled. This toggle is enabled by default in Grafana 11.5 and later.

Reviewer: Same sentiment as my comment above WRT the archived docs and availability info

Author: Removed, thank you.

1. Review error details for any issues that need manual resolution.

## Validate your migration

After you upload a snapshot, use the migration assistant UI to verify that your resources were copied successfully.

The snapshot information panel shows:

- **Total resources**: The number of resources contained in the snapshot.
- **Errors**: The number of errors that occurred while copying resources to Grafana Cloud.
- **Successfully migrated**: The number of resources successfully copied to Grafana Cloud.
- **Uploading to**: The slug of the destination Grafana Cloud stack.

The resource list shows each resource's name, type, and upload status. Use the **Status** column to confirm whether individual resources were copied successfully or require attention.

> ## Snapshots and migration timelines

Reviewer: This heading is fine - I was just thinking of possible alternates and something like "Snapshots and upload performance" might be another to consider

Author: Done

The migration assistant currently supports a subset of all resources available in Grafana. Refer to [Supported resources](#supported-resources) for more details.

When you create a snapshot, the migration assistant makes a copy of all the resources you select and saves them in the snapshot. The snapshot reflects the current state of the resources when the snapshot is built and is stored locally on your instance, ready to be uploaded in the last stage.

> Resources saved in the snapshot are strictly limited to the resources stored within an organization. This is important to note if there are multiple organizations used in your Grafana instance. If you want to migrate multiple organizations, refer to [Migrate multiple organizations](#migrate-multiple-organizations) for more information and guidance.

Reviewer: ```suggestion
Resources saved in the snapshot are strictly limited to the resources stored within an organization. This is important to note if there are multiple organizations used in your Grafana instance. If you want to migrate multiple organizations, refer to [Migrate multiple organizations](#migrate-multiple-organizations).
```

Reviewer: unrelated to your suggestions but this jumped out at me as something that could be tightened up 😄

> Upload time depends primarily on the volume of data you're migrating. The process is optimized for speed and typically takes seconds to minutes rather than hours. In production migrations, the tool has demonstrated 100% success rates with datasets as large as 4,000 objects.

Reviewer: I'm a little iffy on this part. 

I think the first 2 sentences are alright, though "optimized for speed" is a bit vague to me - optimized in what way? I like that it sets expectations which could be useful when planning a migration. I think that track record info in the last sentence is less a technical documentation thing and more marketing-y

Author: I agree. I simplified and removed any speculative/marketing-sounding language.

Most migration operations run asynchronously, so you aren't blocked while a migration is pending and you can perform other tasks on your self-managed instance.

## Rollback and recovery procedures

If issues occur during or after a migration, the following recovery and rollback constraints apply.

### No automated rollback

> There is no tool to automatically undo or roll back a migration. If you need to start fresh, you must manually delete the migrated dashboards, folders, and data sources from your Grafana Cloud instance. On a typical migration path, customers migrate to a new Grafana Cloud stack that has not had previous usage.

Reviewer: ```suggestion
No tool exists to automatically undo or roll back a migration. If you need to start fresh, you must manually delete the migrated dashboards, folders, and data sources from your Grafana Cloud instance. On a typical migration path, you migrate to a new Grafana Cloud stack that doesn't contain existing resources.
```

Reviewer: I'm a bit of 2 minds here.
I think describing the ideal path _can_ be helpful but I also don't want to discourage alternate usage - I know that we've seen folks use the migration assistant to consolidate resources across various instances into a preexisting central stack in the past.
That said, not a huge showstopper of an issue.

- **Single snapshot only**: Only the latest snapshot is available; the system doesn't maintain a history of previous snapshots.
- **One snapshot at a time**: At most one snapshot exists on your self-managed instance at any given time.

### Handling errors and partial migrations

If a migration results in errors, resolve the issues on your self-managed instance, take a new snapshot, and upload it again. Resources on your Grafana Cloud instance are overwritten by the latest snapshot data.

The migration process is non-destructive to the source. Resources on your self-managed instance aren't modified, which serves as an inherent safety measure.

## Disaster recovery and resilience

The Grafana Migration Service (GMS) is designed to recover gracefully from failures during migration.

Resilience to crashes
: GMS uses AWS SQS to distribute work and creates snapshot checkpoints in DynamoDB, allowing it to efficiently pick up where it left off if a crash occurs during a migration.

High availability
: The service runs multiple replicas and depends on highly available AWS services to minimize the likelihood of downtime.

> ## Data privacy and security

Reviewer: This is quite a small section for an H2 (kind of echoing my sentiments in the last section) - it might be worth thinking of consolidating somehow

Author: Good catch. Integrated it into an earlier section.

> Snapshots created on your self-managed instance are encrypted on the filesystem using a key provided by the Grafana Migration Service (GMS). The snapshot is then securely transferred to Grafana Labs cloud infrastructure, where GMS decrypts the data for processing.

Reviewer: ```suggestion
The migration assistant encrypts snapshots on the filesystem of your self-managed instance. The migration assistant transfers the snapshot to the Grafana Labs cloud infrastructure and decrypts the data for processing.
```

Your data sources, including credentials, are migrated securely and seamlessly to your Grafana Cloud instance, so you don't need to find and enter all your data source credentials again.

> Resolve any data source configuration errors on your self-managed instance before you migrate. This helps ensure that error-free data sources appear in Grafana Cloud.

Reviewer: ```suggestion
Resolve any data source configuration errors on your self-managed instance before you migrate. This helps ensure error-free data sources appear in Grafana Cloud.
```

Reviewer: The last sentence might also be phrased better as: "This helps ensure data sources appear in Grafana Cloud without errors." but that's just me spitballing :)

> If your dashboards use data sources on a private network that aren't accessible over the public internet, configure [Private data source connect (PDC)](https://grafana.com/docs/grafana-cloud/connect-externally-hosted/private-data-source-connect/) after migration. PDC lets Grafana Cloud query data sources in your private network without opening inbound access. Confirm that your data source type supports PDC, and configure each applicable data source to connect through PDC after the migration completes.

Reviewer: ```suggestion
If your dashboards use data sources on a private network that aren't accessible over the public internet, configure [Private data source connect (PDC)](https://grafana.com/docs/grafana-cloud/connect-externally-hosted/private-data-source-connect/) after migration. PDC lets Grafana Cloud query data sources in your private network without opening inbound access. Confirm that your data source type supports PDC, and configure each applicable data source to connect through PDC after completing the migration.
```

### Plugins

The migration assistant supports any plugins found in the plugins catalog. As long as the plugin is signed or is a core plugin built into Grafana, you can migrate it. Due to security reasons, Grafana Cloud doesn't support unsigned plugins. If you're using any unsigned private plugins, seek an alternative plugin from the catalog or work on a strategy to deprecate certain functionality from your self-managed instance.

> Only plugins available in the [Grafana plugins catalog](https://grafana.com/grafana/plugins/) can be installed on Grafana Cloud. The migration assistant doesn't support private, custom-built, or third-party plugins that require manual uploading or modifications to Grafana backend files.

Reviewer: ```suggestion
You can only install plugins available in the [Grafana plugins catalog](https://grafana.com/grafana/plugins/) on Grafana Cloud. The migration assistant doesn't support private, custom-built, or third-party plugins that require manual uploading or modifications to Grafana backend files.
```
